### ISSUE TITLE
[Feature] Make OAuth User Specific

### ISSUE THREAD

**Issue by Nikhil-Nandagopal** [appsmithorg/appsmith#3313]
## Summary
There are times when OAuth applications need to be used by a group of users and need to authenticate separately for each individual user. Today OAuth only supports authenticating for a single user and uses that user's credentials for every request.

**Comment by japrogramer** [comment]
Hello, I am trying to get OAuth 2 working as an Authentication system on a per app per user bases like it is done 
with JWT in this tutorial, https://docs.appsmith.com/how-to-guides/how-to-implement-custom-authentication-on-appsmith 

storing the JWT token in the appsmith.store seems simple enough,

Also Noticed that there is an OAuth 2 implementation but it is on a per datasources

https://docs.appsmith.com/core-concepts/connecting-to-data-sources/authentication/oauth2-authentication

It seems logical to store the `Access Token` in the appsmith.store like it is done in the JWT example,

I have two questions, 

1. are API requests in an app done server side on appsmith, API requests like the one made here [JWT](https://docs.appsmith.com/how-to-guides/how-to-implement-custom-authentication-on-appsmith)?

2. can I Include extra code to run server side on an API request ?

**Comment by Nikhil-Nandagopal** [comment]
@japrogramer yes API requests are done server-side with values passed to the server from the client. Extra code cannot be run server-side today. What code are you looking to run server-side?

**Comment by japrogramer** [comment]
@Nikhil-Nandagopal 

I can see that the Data source already has an Authenticated API that supports OAUTh

![image](https://user-images.githubusercontent.com/1387859/137026405-a6b0077e-40db-420b-8718-fb02f1add9f1.png)

Problem is, that the authentication happens when I save the datasource and not on a per visitor basis,

This could be remedied if I could move this button into the app and have the ACCESS-TOKEN header set
on the client side and used on every request to the datastore on each visitor so that each visitor needs to
authenticate with the datastore first before using the APIs that are assigned to that store. 

![image](https://user-images.githubusercontent.com/1387859/137026658-db9a9727-a27a-4ec1-ab33-b984b266054f.png)

And it would be even better if I could then wrap the OAuth authentication around a database so that I don't have to pass the credentials for the database to the app. Instead let me worry about checking the token.

**Comment by Nikhil-Nandagopal** [comment]
@japrogramer thanks! Yes user-specific authentication does make sense and is on our roadmap

**Comment by nidhi-nair** [comment]
Closing in favour of #10434 since there are more implementation details there.

**Comment by Nikhil-Nandagopal** [comment]
@nidhi-nair I'm keeping this one open since we get most of our upvotes on this issue just for community tracking

**Comment by hreeder** [comment]
a 👍  from me on this request too. My use case is connecting to larger scale backend systems which already have constraints in place on what can be requested based on the current user. In the current implementation where OAuth is tied to the developer's identity, this means that our backend would only show results which the developer can access.

**Comment by Nikhil-Nandagopal** [comment]
Hi @hreeder 👋 
Actually, this feature is primarily for SAAS integrations like google sheets. If you're working with a backend and want to authenticate your users based on some OAuth / OIDC / SAML mechanism, you can reach out to us for our commercial edition which supports this today!

**Comment by hreeder** [comment]
Hey @Nikhil-Nandagopal
This feature makes sense - are you referring to authenticating users to Appsmith with OAuth/OIDC/SAML which is already implemented in enterprise, or API requests using per-user OAuth to a backend API (I say backend system meaning other SaaS platforms my team already makes use of here)

**Comment by Nikhil-Nandagopal** [comment]
@hreeder Yes I meant the SAML / OIDC one. There you can then forward the JWT token as part of your own API calls if you are using this in your own custom backend. If you are using other SAAS platforms. Could you tell me which ones are you using that need this feature?

**Comment by hreeder** [comment]
ServiceNow is the main platform that comes to mind where having this functionality would be helpful. That said, interacting with GitHub would also be an interesting avenue for this feature as well.

**Comment by Nikhil-Nandagopal** [comment]
@hreeder That's really interesting! Doesn't service now have it's own internal tool platform. I'd love to learn about why you're integrating ServiceNow with Appsmith. Would you be available for a quick chat? This is my Calendly below
https://calendly.com/appsmith-nikhil/30min?back=1&month=2022-02

**Comment by hreeder** [comment]
Dropped in a slot via Calendly, looking forward to talking with you :)

**Comment by pranavkanade** [comment]
A user describes their use case here - https://community.appsmith.com/t/google-api-calls-from-user-in-stead-of-from-developer/1826

**Comment by antorappai** [comment]
Hi, i just came across this post, i have a use case where i need a response from an API call and would need to follow up with a create Google Doc and Update Google Doc under that user's account. As of now it creates it under my account. 

I suppose the only way to overcome this would be to create a whole JS Object script and recreate the whole flow, i can not even comprehend this. 

Could i import a library to do this ? Could you point me ? 

Alternatively i could add the logged in user's email address as an editor for the Doc that has been created in my account, is there a way i can retrieve the logged in users email address ?

**Comment by Nikhil-Nandagopal** [comment]
@antorappai yes currently we do not support that use case

**Comment by antorappai** [comment]
i could however do this 
Alternatively i could add the logged in user's email address as an editor for the Doc that has been created in my account, is there a way i can retrieve the logged in users email address ?

and retrive the current logged in user

**Comment by antorappai** [comment]
ok i see the reason is appsmith wont let us run backend scripts as of now, sorry some novice questions coming from my side

**Comment by DeLaiT** [comment]
Is it possible to prioritize this feature? 
Unfortunatelly the current way of how the Oauth implementation works makes Appsmith unusable for my team's usecase and makes alternative services (ToolJet, Reetol) more convincing.
